File: src/main/java/de/mzte/multiutil/module/ModuleLoader.java
package de.mzte.multiutil.module;

import org.reflections.Reflections;
import org.reflections.scanners.SubTypesScanner;

import java.util.HashMap;
import java.util.Map;
import java.util.NoSuchElementException;

public class ModuleLoader {
    private final String pkg;
    private final Map<String, IModule> modules = new HashMap<>();

    public ModuleLoader(String pkg) {
        this.pkg = pkg;
        init();
    }

    private void init() {
        Reflections refs = new Reflections(pkg, new SubTypesScanner(false));
        refs.getSubTypesOf(IModule.class).stream()
                .filter(c -> c.isAnnotationPresent(LoadModule.class))
                .map(c -> {
                    try {
                        return (IModule) c.newInstance();
                    } catch(InstantiationException | IllegalAccessException e) {
                        e.printStackTrace();
                    }
                    return null;
                }).peek(m -> System.out.println("Loading Module " + m.getID()))
                .forEach(this::addModule);
    }

    public void addModule(IModule module) {
        addModule(module, module.getID());
    }

    public void addModule(IModule module, String id) {
        modules.put(id, module);
    }

    public void runModule(String id) {
        modules.computeIfAbsent(id,
                x -> {
                    throw new NoSuchElementException("Command " + id + " Not Found");
                }).run();
    }

    public void runModuleAsync(String id) {
        (new Thread(() -> runModule(id))).start();
    }

    public Map<String, IModule> getModules() {
        return new HashMap<>(modules);
    }
}

File: src/main/kotlin/de/mzte/multiutil/module/ModuleLoader.kt
package de.mzte.multiutil.module

import org.reflections.Reflections
import org.reflections.scanners.SubTypesScanner
import kotlin.reflect.KClass
import kotlin.reflect.full.createInstance

class ModuleLoader(val pkg: String) {
    val modules = mutableMapOf<String, IModule>()

    init {
        loadModules()
    }

    @Suppress("UNCHECKED_CAST")
    private fun loadModules() {
        val refs = Reflections(pkg, SubTypesScanner(false))
        refs.getSubTypesOf(IModule::class.java)
            .map {it.kotlin}
            .filter {it.annotations.any {a -> a.annotationClass == LoadModule::class}}
            .map {it.objectInstance ?: it}
            .forEach {if(it is IModule) addModule(it) else addClass(it as KClass<out IModule>)}
    }

    @JvmOverloads
    fun addModule(module: IModule, id: String = module.id) {
        modules[id] = module
    }

    fun addClass(clazz: KClass<out IModule>) = addModule(clazz.createInstance())

    fun runModule(id: String) = modules.getOrElse(id) {throw NoSuchElementException("Command $id not found")}.run()
}